src/lesson-1-9-demo/SignupForm.jsx

```diff
@@ -1,23 +1,165 @@
 import { useEffect, useState } from "react";
 import "./signup-form.css";
 
+const createUser = async (userData) => {
+  try {
+    const response = await fetch("http://localhost:3001/api/users", {
+      headers: {
+        Accept: "application/json",
+        "Content-Type": "application/json",
+      },
+      method: "POST",
+      body: JSON.stringify(userData),
+    });
+
+    if (response.ok) {
+      return response.json();
+    }
+
+    throw Error("Failed to create user");
+  } catch (error) {
+    console.error({ error });
+    throw error;
+  }
+};
+
+const initialFormState = {
+  username: "",
+  email: "",
+  password: "",
+  passwordConfirmation: "",
+};
+
 const SignupForm = () => {
+  const [formState, setFormState] = useState(initialFormState);
+  const [formErrors, setFormErrors] = useState({});
+
+  const validations = [
+    {
+      key: "username",
+      isValid: (formState) => formState.username.trim().length > 0,
+      errorMessage: "username cannot be blank",
+    },
+    {
+      key: "email",
+      isValid: (formState) =>
+        formState.email.length > 0 &&
+        /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/.test(formState.email),
+      errorMessage: "email is not valid",
+    },
+    {
+      key: "password",
+      isValid: (formState) => {
+        return (
+          formState.password.length > 12 &&
+          /[a-z]/.test(formState.password) &&
+          /[A-Z]/.test(formState.password) &&
+          /[0-9]/.test(formState.password) &&
+          /[!@#$%^&*]/.test(formState.password)
+        );
+      },
+      errorMessage: "password is not valid",
+    },
+    {
+      key: "passwordConfirmation",
+      isValid: (formState) => {
+        console.log(formState.password, formState.passwordConfirmation);
+        return formState.password === formState.passwordConfirmation;
+      },
+      errorMessage: "password confirmation does not match password",
+    },
+  ];
+  const validateFormState = (formState) => {
+    const errors = {};
+    validations.forEach((validation) => {
+      const isValid = validation.isValid(formState);
```

Comment:
we run through all of the `validations` and call the `isValid` function on the `formState`

if any `isValid` call is invalid, then we update the `errors` object to store the results of the validation.